### PR TITLE
spi: stm32: spi_ll_stm32: add MSSI and MIDI support

### PR DESCRIPTION
**Pull request by mustafaabdullahk**
Hi everyone

## Adjustable Signal Flow Timing Support

STM32H7xx series mcu does support that feaute.

image

Several parameters can be used to optionally adjust the timing of the master transaction flow. Signal timings for the master node can be adjusted when needed. This is the case when a slave node needs a longer time to wake up from sleep mode following the setup of the SS active edge or when the slave is not able to handle a data flow that is too fast. Up to 15 additional serial clock signal periods delays can be inserted by MIDI or MSSI parameters. Data frames can be optionally interleaved by SS pulses.

Signed-off-by: Mustafa Abdullah Kus [mustafa.kus@sparsetechnology.com](mailto:mustafa.kus@sparsetechnology.com)

### CODE REVIEW

File: dts/bindings/spi/st,stm32h7-spi.yaml [hunk 1]
```diff
@@ -12,3 +12,18 @@ description: |
 compatible: "st,stm32h7-spi"
 
 include: st,stm32-spi-common.yaml
+
+properties:
+  midi-clock:
+    type: int
+    default: 0
+    description: |
+      Only for STM32H7, (Master Inter-Data Idleness) minimum
```
Comment (erwango) [inline dts/bindings/spi/st,stm32h7-spi.yaml:21]: Restriction `Only for STM32H7,` is misleading. This property is made available on `st,stm32h7-spi` compatible, hence will be available for compatible SoCs: H5/H7/MP1/U5/WBA. This is perfectly fine and sufficient

File: drivers/spi/spi_ll_stm32.c [hunk 2]
```diff
@@ -618,6 +618,49 @@ static int spi_stm32_configure(const struct device *dev,
 		LL_SPI_SetDataWidth(spi, LL_SPI_DATAWIDTH_16BIT);
 	}
 
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi)
+	const uint32_t mssi_cycles[] = {
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c:622]: These table are an expensive way to configure this. Also, it seems that values encoding looks fairly simple.
What about using directly dt values w/o these tables ?

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this addition!
Looks globally good (Device Tree usage was the tricky part) but I'd like to explore some optimization possibility.

File: drivers/spi/spi_ll_stm32.c (hunk 2 above)
Comment (mustafaabdullahk) [reply drivers/spi/spi_ll_stm32.c]: i totally aggree with you but following values shifting: 
image
so i didn't bit shift at the start but update for your suggestion

File: dts/bindings/spi/st,stm32h7-spi.yaml (hunk 1 above)
Comment (mustafaabdullahk) [reply dts/bindings/spi/st,stm32h7-spi.yaml]: good point, i did update document :)

File: drivers/spi/spi_ll_stm32.c [hunk 3]
```diff
@@ -1160,6 +1165,12 @@ static const struct spi_stm32_config spi_stm32_cfg_##id = {		\
 	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32_spi_subghz),	\
 		(.use_subghzspi_nss =					\
 			DT_INST_PROP_OR(id, use_subghzspi_nss, false),))\
+	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi),	\
+		(.midi_clocks =					\
+			DT_INST_PROP_OR(id, midi_clock, 0),))\
+	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi),	\
+		(.mssi_clocks =					\
+			DT_INST_PROP_OR(id, mssi_clock, 0),))\
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c:1173]: Please mind indentation

```suggestion
	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi),		\
		(.midi_clocks =						\
			DT_INST_PROP_OR(id, midi_clock, 0),))		\
	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi),		\
		(.mssi_clocks =						\
			DT_INST_PROP_OR(id, mssi_clock, 0),))		\
```

File: drivers/spi/spi_ll_stm32.c [hunk 4]
```diff
@@ -1160,6 +1165,12 @@ static const struct spi_stm32_config spi_stm32_cfg_##id = {		\
 	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32_spi_subghz),	\
 		(.use_subghzspi_nss =					\
 			DT_INST_PROP_OR(id, use_subghzspi_nss, false),))\
+	IF_ENABLED(DT_HAS_COMPAT_STATUS_OKAY(st_stm32h7_spi),	\
+		(.midi_clocks =					\
+			DT_INST_PROP_OR(id, midi_clock, 0),))\
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c:1170]: As you provide a default value, prop will automatically be generated with default value (0) if not filled by the user.
Then you don't need to use `DT_INST_PROP_OR` and can just use `DT_INST_PROP(id, midi_clock)`.
Same for the other prop.

Note: It should actually be the same for `use_subghzspi_nss` as boolean properties are false if not present, no need for the `_OR` variant. I'll create an issue to get a clean up done.

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for the update.
2 minor fixes that I've just seen and we're good to go.

**Review by erwango — APPROVED** [review]

File: drivers/spi/spi_ll_stm32.c (hunk 3 above)
Comment (mustafaabdullahk) [reply drivers/spi/spi_ll_stm32.c]: my tab size value did keep as 4, i did update and solve. :)

File: drivers/spi/spi_ll_stm32.c (hunk 4 above)
Comment (mustafaabdullahk) [reply drivers/spi/spi_ll_stm32.c]: i did update `DT_INST_PROP_OR`,  may i solve use_subghzspi_nss issue with another PR if you want assign to me.

Comment (erwango) [reply drivers/spi/spi_ll_stm32.c]: I've created https://github.com/zephyrproject-rtos/zephyr/issues/67651 to fix this globally. You're welcome to have a stab at it!

**Review by gautierg-st — APPROVED** [review]

**Pull request merged**